floodmodeller_api/units/_base.py

```diff
@@ -99,3 +109,144 @@ def __eq__(self, other, return_diff=False):
+                self.time_varrule_data = varrule_data
+
+    def _write_rules(self, block):
+        # ADD RULES
+        block.append("RULES")
+        self.nrules = len(self.rules)
+        block.append(
+            f"{join_n_char_ljust(10, self.nrules)}{join_10_char(self.rule_sample_time)}{join_n_char_ljust(10, self.timeunit, self.extendmethod)}"
+        )
+        for rule in self.rules:
+            try:
+                if rule["logic"].strip().upper().endswith(("END", "ENDIF")):
+                    block.append(rule["name"])
+                    block.extend(rule["logic"].split("\n"))
+                else:
+                    raise Exception(
+                        f"You have tried to add a new rule ({rule['name']}) to {self.name}.rules, the logic of which does not end in either END or ENDIF"
+                    )
+            except Exception as e:
+                self._handle_exception(e, when="add a new rule")
+
+        # ADD TIME RULE DATA SET
+        block.append("TIME RULE DATA SET")
+        block.append(join_10_char(len(self.time_rule_data)))
+        time_rule_data = [
+            f"{join_10_char(t)}{o_r:<10}" for t, o_r in self.time_rule_data.items()
+        ]
+        block.extend(time_rule_data)
+
+        # ADD VARRULES (IF THEY ARE THERE)
+        if self.has_varrules:
+            block.append("VARRULES")
+            self.nvarrules = len(self.varrules)
+            block.append(
+                f"{join_n_char_ljust(10, self.nvarrules)}{join_10_char(self.varrule_sample_time)}{join_n_char_ljust(10, self.timeunit, self.extendmethod)}"
+            )
+            for varrule in self.varrules:
+                block.append(varrule["name"])
+                block.extend(varrule["logic"].split("\n"))
+
+            # ADD TIME VARRULE DATA SET
+            block.append("TIME RULE DATA SET")
+            block.append(join_10_char(len(self.time_varrule_data)))
+            time_varrule_data = [
+                f"{join_10_char(t)}{o_r:<10}" for t, o_r in self.time_rule_data.items()
+            ]
+            block.extend(time_varrule_data)
+
+        return block
+
+    def _get_logical_rules(self, nrules, block, rule_row):
+        rules = []
+        rules_recorded = 0
+        rule_logic = []
+        rule_dict = {}
+        nl = "\n"
+        while rules_recorded < nrules:
+            if block[rule_row].strip().upper().endswith(("END", "ENDIF")):
+                rule_logic.append(block[rule_row])
+                rule_dict["logic"] = f"{nl.join(rule_logic)}"
+                rules.append(rule_dict)
+                rule_logic = []
+                rule_dict = {}
+                rules_recorded += 1
+            elif len(rule_dict) == 0:
+                rule_dict = {"name": block[rule_row].strip()}
+            else:
+                rule_logic.append(block[rule_row])
+            rule_row += 1
+
+        self._last_rule_row = rule_row
+
+        return rules
+
+    def _handle_exception(self, err, when):
```

Comment:
I may have miscommunicated here - we don't want to add this method again here. Exceptions on writing units will be caught by the existing exception handling on FMFile, so this doesn't need to be added here.

Reply:
Okay I understand, I thought the validation and parameters was for something separate so I was handling the exception by doing it a different way - I will use the validation and parameters for it now.